third_party/xla_client/pjrt_computation_client.cc

```diff
@@ -145,26 +175,49 @@ std::vector<ComputationClient::ComputationPtr> PjRtComputationClient::Compile(
   std::vector<ComputationClient::ComputationPtr> computations;
 
   for (auto& instance : instances) {
-    PjRtDevice* pjrt_device = StringToPjRtDevice(instance.compilation_device);
-    xla::ProgramShape program_shape =
-        instance.computation.GetProgramShape().ValueOrDie();
     xla::CompileOptions compile_options;
-    xla::DeviceAssignment device_assignment(client_->device_count(), 1);
-    device_assignment.FillIota(0);
-    compile_options.executable_build_options.set_device_assignment(
-        device_assignment);
-    // TODO(wcromar): set compile_options.argument_layouts, enable strict shapes
-    compile_options.executable_build_options.set_num_partitions(1);
-    compile_options.executable_build_options.set_num_replicas(
-        client_->device_count());
-    compile_options.parameter_is_tupled_arguments =
-        instance.parameter_is_tupled_arguments;
+    if (instance.is_sharded) {
+      // TODO(yeounoh) multi-host, multi-slice configurations
+      compile_options.executable_build_options.set_use_spmd_partitioning(true);
+      compile_options.executable_build_options.set_num_partitions(
+          client_->device_count());
+      compile_options.executable_build_options.set_num_replicas(1);
+      compile_options.parameter_is_tupled_arguments =
+          instance.parameter_is_tupled_arguments;
+
+      // TODO(244391366) verify this is correct for the collectives ops
+      xla::DeviceAssignment device_assignment(1, client_->device_count());
+      device_assignment.FillIota(0);
+      compile_options.executable_build_options.set_device_assignment(
+          device_assignment);
+    } else {
+      // TODO(wcromar): set compile_options.argument_layouts, enable strict
+      // shapes
+      compile_options.executable_build_options.set_num_partitions(1);
+      compile_options.executable_build_options.set_num_replicas(
+          client_->device_count());
+      compile_options.parameter_is_tupled_arguments =
+          instance.parameter_is_tupled_arguments;
+
+      xla::DeviceAssignment device_assignment(client_->device_count(), 1);
+      device_assignment.FillIota(0);
+      compile_options.executable_build_options.set_device_assignment(
+          device_assignment);
+    }
+
+    PjRtDevice* pjrt_device = StringToPjRtDevice(instance.compilation_device);
     std::unique_ptr<xla::PjRtExecutable> executable =
-        client_->Compile(instance.computation, compile_options).ValueOrDie();
+        ConsumeValue(client_->Compile(instance.computation, compile_options));
```

Comment:
Is there any difference in behavior between `ConsumeValue` and `ValueOrDie`? It seems to use a deprecated reference to `ValueOrDie` anyway:

https://github.com/pytorch/xla/blob/b7f54864a0933d6401a023e12ec6ccac1982e2d9/third_party/xla_client/debug_macros.h#L25-L29

`ConsumeValueOrDie`: 
 https://github.com/tensorflow/tensorflow/blob/57d35fb585de65a34cd6871b8ca3a6501183f98a/tensorflow/core/platform/statusor.h#L234-L235

Reply:
There is an important distinction between `ConsumeValueOrDie` and `ValueOrDie` -- the former deletes the old copy, which we don't need anymore. It's better to "consume".